# Analyzing Data Sets with Pandas #

The goal is to use Pandas to extract useful information from a data set about average crop yields from a list of countries in the OECD (the Organization for Economic Cooperation and Development).  In particular, we will be stepping through some Pandas commands that will allow us to explore the data and answer the question:

**BIG QUESTION:** What are the average crop yields for rice in Vietnam for the most recent ten years?  (Note:  This is a slighly different question from what I asked in class, becuase I forgot that the data estimates future results.) 

In [1]:
import pandas as pd

In [5]:
mydata = pd.read_csv('cropyields.csv')

In [6]:
mydata.head(5)

LOCATION  INDICATOR SUBJECT   MEASURE FREQUENCY  TIME  Value  Flag Codes
0      AUS  CROPYIELD    RICE  TONNE_HA         A  1990  8.315         NaN
1      AUS  CROPYIELD    RICE  TONNE_HA         A  1991  8.395         NaN
2      AUS  CROPYIELD    RICE  TONNE_HA         A  1992  8.094         NaN
3      AUS  CROPYIELD    RICE  TONNE_HA         A  1993  8.336         NaN
4      AUS  CROPYIELD    RICE  TONNE_HA         A  1994  8.538         NaN

**Question:** What are the countries represented?  What crops are represented?  What years are represented? What are the measures used?

In [7]:
mydata["LOCATION"].unique()

array(['AUS', 'CAN', 'JPN', 'KOR', 'MEX', 'NZL', 'TUR', 'USA', 'ARG',
       'BRA', 'CHL', 'CHN', 'COL', 'EGY', 'ETH', 'IND', 'IDN', 'IRN',
       'KAZ', 'MYS', 'NGA', 'PAK', 'PRY', 'PER', 'PHL', 'RUS', 'SAU',
       'ZAF', 'THA', 'UKR', 'VNM', 'WLD', 'OECD', 'BRICS', 'NOR', 'CHE',
       'ISR', 'GBR'], dtype=object)

In [8]:
mydata["SUBJECT"].unique()

array(['RICE', 'WHEAT', 'MAIZE', 'SOYBEAN'], dtype=object)

In [9]:
mydata["TIME"].unique()

array([1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000,
       2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011,
       2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022,
       2023, 2024, 2025, 2026, 2027, 2028, 2029, 2030])

In [10]:
mydata["MEASURE"].unique()

array(['TONNE_HA', 'THND_TONNE', 'THND_HA'], dtype=object)

What data represents Vietnam? (These have LOCATION = VNM.)  What data represents rice?  (These have SUBJECT = RICE.)

In [12]:
mydata[(mydata["LOCATION"] == "VNM") & (mydata["SUBJECT"] == "RICE")]

LOCATION  INDICATOR SUBJECT     MEASURE FREQUENCY  TIME      Value  \
2419      VNM  CROPYIELD    RICE    TONNE_HA         A  1990      2.122   
2420      VNM  CROPYIELD    RICE    TONNE_HA         A  1991      2.077   
2421      VNM  CROPYIELD    RICE    TONNE_HA         A  1992      2.224   
2422      VNM  CROPYIELD    RICE    TONNE_HA         A  1993      2.322   
2423      VNM  CROPYIELD    RICE    TONNE_HA         A  1994      2.378   
...       ...        ...     ...         ...       ...   ...        ...   
6371      VNM  CROPYIELD    RICE  THND_TONNE         A  2026  30695.302   
6372      VNM  CROPYIELD    RICE  THND_TONNE         A  2027  31217.765   
6373      VNM  CROPYIELD    RICE  THND_TONNE         A  2028  31739.779   
6374      VNM  CROPYIELD    RICE  THND_TONNE         A  2029  32282.995   
6375      VNM  CROPYIELD    RICE  THND_TONNE         A  2030  32823.537   

      Flag Codes  
2419         NaN  
2420         NaN  
2421         NaN  
2422         NaN  
2423         NaN  
...          ...  
6371         NaN  
6372         NaN  
6373         NaN  
6374         NaN  
6375         NaN  

[123 rows x 8 columns]

Here is the data for the most recent 10 years. 

In [13]:
mydata[(mydata["LOCATION"] == "VNM") & (mydata["SUBJECT"] == "RICE") & (mydata['TIME'] > 2014) & (mydata['TIME'] < 2026)]

LOCATION  INDICATOR SUBJECT     MEASURE FREQUENCY  TIME      Value  \
2444      VNM  CROPYIELD    RICE    TONNE_HA         A  2015      3.744   
2445      VNM  CROPYIELD    RICE    TONNE_HA         A  2016      3.626   
2446      VNM  CROPYIELD    RICE    TONNE_HA         A  2017      3.605   
2447      VNM  CROPYIELD    RICE    TONNE_HA         A  2018      3.782   
2448      VNM  CROPYIELD    RICE    TONNE_HA         A  2019      3.786   
2449      VNM  CROPYIELD    RICE    TONNE_HA         A  2020      3.813   
2450      VNM  CROPYIELD    RICE    TONNE_HA         A  2021      3.877   
2451      VNM  CROPYIELD    RICE    TONNE_HA         A  2022      3.941   
2452      VNM  CROPYIELD    RICE    TONNE_HA         A  2023      4.009   
2453      VNM  CROPYIELD    RICE    TONNE_HA         A  2024      4.076   
2454      VNM  CROPYIELD    RICE    TONNE_HA         A  2025      4.144   
3654      VNM  CROPYIELD    RICE     THND_HA         A  2015   7828.000   
3655      VNM  CROPYIELD    RICE     THND_HA         A  2016   7737.000   
3656      VNM  CROPYIELD    RICE     THND_HA         A  2017   7705.200   
3657      VNM  CROPYIELD    RICE     THND_HA         A  2018   7570.900   
3658      VNM  CROPYIELD    RICE     THND_HA         A  2019   7472.000   
3659      VNM  CROPYIELD    RICE     THND_HA         A  2020   7277.800   
3660      VNM  CROPYIELD    RICE     THND_HA         A  2021   7282.237   
3661      VNM  CROPYIELD    RICE     THND_HA         A  2022   7278.162   
3662      VNM  CROPYIELD    RICE     THND_HA         A  2023   7286.766   
3663      VNM  CROPYIELD    RICE     THND_HA         A  2024   7286.854   
3664      VNM  CROPYIELD    RICE     THND_HA         A  2025   7287.703   
6360      VNM  CROPYIELD    RICE  THND_TONNE         A  2015  29309.150   
6361      VNM  CROPYIELD    RICE  THND_TONNE         A  2016  28057.315   
6362      VNM  CROPYIELD    RICE  THND_TONNE         A  2017  27780.285   
6363      VNM  CROPYIELD    RICE  THND_TONNE         A  2018  28629.900   
6364      VNM  CROPYIELD    RICE  THND_TONNE         A  2019  28285.400   
6365      VNM  CROPYIELD    RICE  THND_TONNE         A  2020  27747.720   
6366      VNM  CROPYIELD    RICE  THND_TONNE         A  2021  28229.694   
6367      VNM  CROPYIELD    RICE  THND_TONNE         A  2022  28685.405   
6368      VNM  CROPYIELD    RICE  THND_TONNE         A  2023  29215.288   
6369      VNM  CROPYIELD    RICE  THND_TONNE         A  2024  29698.854   
6370      VNM  CROPYIELD    RICE  THND_TONNE         A  2025  30203.162   

      Flag Codes  
2444         NaN  
2445         NaN  
2446         NaN  
2447         NaN  
2448         NaN  
2449         NaN  
2450         NaN  
2451         NaN  
2452         NaN  
2453         NaN  
2454         NaN  
3654         NaN  
3655         NaN  
3656         NaN  
3657         NaN  
3658         NaN  
3659         NaN  
3660         NaN  
3661         NaN  
3662         NaN  
3663         NaN  
3664         NaN  
6360         NaN  
6361         NaN  
6362         NaN  
6363         NaN  
6364         NaN  
6365         NaN  
6366         NaN  
6367         NaN  
6368         NaN  
6369         NaN  
6370         NaN

**Note:** There are *three* measurements going on here--one in tonnes per hectare (TONNE_HA), another in thousands on hectares (THND_HA), and a third in thousands on tonnes (THND_TONNE).  These three measurements give:

+ the crop yield per hectare (TONNE_HA)
+ the number of hectares devoted to this crop (THND_HA)
+ the number of tonnes of rice produces (THND_TONNE)

**We are interested in the first one.**

**Moral of the story:**  Know your data.

In [21]:
mydata[(mydata["LOCATION"] == "VNM") & (mydata["SUBJECT"] == "RICE") & (mydata['TIME'] > 2014) & (mydata['TIME'] < 2026) & (mydata['MEASURE'] =='TONNE_HA')]

LOCATION  INDICATOR SUBJECT   MEASURE FREQUENCY  TIME  Value  Flag Codes
2444      VNM  CROPYIELD    RICE  TONNE_HA         A  2015  3.744         NaN
2445      VNM  CROPYIELD    RICE  TONNE_HA         A  2016  3.626         NaN
2446      VNM  CROPYIELD    RICE  TONNE_HA         A  2017  3.605         NaN
2447      VNM  CROPYIELD    RICE  TONNE_HA         A  2018  3.782         NaN
2448      VNM  CROPYIELD    RICE  TONNE_HA         A  2019  3.786         NaN
2449      VNM  CROPYIELD    RICE  TONNE_HA         A  2020  3.813         NaN
2450      VNM  CROPYIELD    RICE  TONNE_HA         A  2021  3.877         NaN
2451      VNM  CROPYIELD    RICE  TONNE_HA         A  2022  3.941         NaN
2452      VNM  CROPYIELD    RICE  TONNE_HA         A  2023  4.009         NaN
2453      VNM  CROPYIELD    RICE  TONNE_HA         A  2024  4.076         NaN
2454      VNM  CROPYIELD    RICE  TONNE_HA         A  2025  4.144         NaN

Since this is the data we want, we will create a new dataset to hold this data.

In [16]:
newdata = mydata[(mydata["LOCATION"] == "VNM") & (mydata["SUBJECT"] == "RICE") & (mydata['TIME'] > 2014) & (mydata['TIME'] < 2026) & (mydata['MEASURE'] =='TONNE_HA')]

In [17]:
newdata.head(5)

LOCATION  INDICATOR SUBJECT   MEASURE FREQUENCY  TIME  Value  Flag Codes
2444      VNM  CROPYIELD    RICE  TONNE_HA         A  2015  3.744         NaN
2445      VNM  CROPYIELD    RICE  TONNE_HA         A  2016  3.626         NaN
2446      VNM  CROPYIELD    RICE  TONNE_HA         A  2017  3.605         NaN
2447      VNM  CROPYIELD    RICE  TONNE_HA         A  2018  3.782         NaN
2448      VNM  CROPYIELD    RICE  TONNE_HA         A  2019  3.786         NaN

Now we answer our question:

In [19]:
newdata['Value'].mean()

np.float64(3.8548181818181817)

**Answer:** The average yield for rice in Vietnam is 3.85 tonnes per hectare.

Note that this information seems accurate given that the 'Value' for each row seems to be about 3.7 or so tonnes per hectare.

In [27]:
mydata[(mydata["LOCATION"] == "USA") & (mydata["SUBJECT"] == "RICE") & (mydata['TIME'] > 1989) & (mydata['TIME'] < 2021) & (mydata['MEASURE'] =='THND_HA')]

LOCATION  INDICATOR SUBJECT  MEASURE FREQUENCY  TIME     Value  \
7811      USA  CROPYIELD    RICE  THND_HA         A  1990  1142.431   
7812      USA  CROPYIELD    RICE  THND_HA         A  1991  1125.435   
7813      USA  CROPYIELD    RICE  THND_HA         A  1992  1267.480   
7814      USA  CROPYIELD    RICE  THND_HA         A  1993  1146.478   
7815      USA  CROPYIELD    RICE  THND_HA         A  1994  1341.942   
7816      USA  CROPYIELD    RICE  THND_HA         A  1995  1251.697   
7817      USA  CROPYIELD    RICE  THND_HA         A  1996  1134.742   
7818      USA  CROPYIELD    RICE  THND_HA         A  1997  1255.744   
7819      USA  CROPYIELD    RICE  THND_HA         A  1998  1318.066   
7820      USA  CROPYIELD    RICE  THND_HA         A  1999  1421.261   
7821      USA  CROPYIELD    RICE  THND_HA         A  2000  1229.844   
7822      USA  CROPYIELD    RICE  THND_HA         A  2001  1341.133   
7823      USA  CROPYIELD    RICE  THND_HA         A  2002  1297.831   
7824      USA  CROPYIELD    RICE  THND_HA         A  2003  1212.847   
7825      USA  CROPYIELD    RICE  THND_HA         A  2004  1345.180   
7826      USA  CROPYIELD    RICE  THND_HA         A  2005  1359.748   
7827      USA  CROPYIELD    RICE  THND_HA         A  2006  1149.311   
7828      USA  CROPYIELD    RICE  THND_HA         A  2007  1112.889   
7829      USA  CROPYIELD    RICE  THND_HA         A  2008  1205.967   
7830      USA  CROPYIELD    RICE  THND_HA         A  2009  1254.530   
7831      USA  CROPYIELD    RICE  THND_HA         A  2010  1464.967   
7832      USA  CROPYIELD    RICE  THND_HA         A  2011  1060.280   
7833      USA  CROPYIELD    RICE  THND_HA         A  2012  1084.561   
7834      USA  CROPYIELD    RICE  THND_HA         A  2013   999.577   
7835      USA  CROPYIELD    RICE  THND_HA         A  2014  1185.733   
7836      USA  CROPYIELD    RICE  THND_HA         A  2015  1048.139   
7837      USA  CROPYIELD    RICE  THND_HA         A  2016  1254.530   
7838      USA  CROPYIELD    RICE  THND_HA         A  2017   959.108   
7839      USA  CROPYIELD    RICE  THND_HA         A  2018  1177.639   
7840      USA  CROPYIELD    RICE  THND_HA         A  2019  1031.952   
7841      USA  CROPYIELD    RICE  THND_HA         A  2020  1230.249   

      Flag Codes  
7811         NaN  
7812         NaN  
7813         NaN  
7814         NaN  
7815         NaN  
7816         NaN  
7817         NaN  
7818         NaN  
7819         NaN  
7820         NaN  
7821         NaN  
7822         NaN  
7823         NaN  
7824         NaN  
7825         NaN  
7826         NaN  
7827         NaN  
7828         NaN  
7829         NaN  
7830         NaN  
7831         NaN  
7832         NaN  
7833         NaN  
7834         NaN  
7835         NaN  
7836         NaN  
7837         NaN  
7838         NaN  
7839         NaN  
7840         NaN  
7841         NaN